File: pkg/video/hls/segmenter.go
package hls

import (
	"bytes"
	"nvr/pkg/video/gortsplib"
	"nvr/pkg/video/gortsplib/pkg/h264"
	"time"
)

func partDurationIsCompatible(partDuration time.Duration, sampleDuration time.Duration) bool {
	if sampleDuration > partDuration {
		return false
	}

	f := (partDuration / sampleDuration)
	if (partDuration % sampleDuration) != 0 {
		f++
	}
	f *= sampleDuration

	return partDuration > ((f * 85) / 100)
}

func partDurationIsCompatibleWithAll(partDuration time.Duration, sampleDurations map[time.Duration]struct{}) bool {
	for sd := range sampleDurations {
		if !partDurationIsCompatible(partDuration, sd) {
			return false
		}
	}
	return true
}

func findCompatiblePartDuration(
	minPartDuration time.Duration,
	sampleDurations map[time.Duration]struct{},
) time.Duration {
	i := minPartDuration
	for ; i < 5*time.Second; i += 5 * time.Millisecond {
		if partDurationIsCompatibleWithAll(i, sampleDurations) {
			break
		}
	}
	return i
}

type segmenter struct {
	segmentDuration    time.Duration
	partDuration       time.Duration
	segmentMaxSize     uint64
	videoTrack         *gortsplib.TrackH264
	audioTrack         *gortsplib.TrackMPEG4Audio
	onSegmentFinalized func(*Segment)
	onPartFinalized    func(*MuxerPart)

	startDTS              time.Duration
	muxerStartTime        int64
	videoFirstIDRReceived bool
	videoDTSExtractor     *h264.DTSExtractor
	nextSegmentID         uint64
	videoSPS              []byte
	currentSegment        *Segment
	nextPartID            uint64
	nextVideoSample       *VideoSample
	nextAudioSample       *AudioSample
	firstSegmentFinalized bool
	sampleDurations       map[time.Duration]struct{}
	adjustedPartDuration  time.Duration
}

func newSegmenter(
	muxerStartTime int64,
	segmentDuration time.Duration,
	partDuration time.Duration,
	segmentMaxSize uint64,
	videoTrack *gortsplib.TrackH264,
	audioTrack *gortsplib.TrackMPEG4Audio,
	onSegmentFinalized func(*Segment),
	onPartFinalized func(*MuxerPart),
) *segmenter {
	return &segmenter{
		segmentDuration:    segmentDuration,
		partDuration:       partDuration,
		segmentMaxSize:     segmentMaxSize,
		videoTrack:         videoTrack,
		audioTrack:         audioTrack,
		onSegmentFinalized: onSegmentFinalized,
		onPartFinalized:    onPartFinalized,
		muxerStartTime:     muxerStartTime,
		nextSegmentID:      7, // Required by iOS.
		sampleDurations:    make(map[time.Duration]struct{}),
	}
}

func (m *segmenter) genSegmentID() uint64 {
	id := m.nextSegmentID
	m.nextSegmentID++
	return id
}

func (m *segmenter) genPartID() uint64 {
	id := m.nextPartID
	m.nextPartID++
	return id
}

// iPhone iOS fails if part durations are less than 85% of maximum part duration.
// find a part duration that is compatible with all received sample durations.
func (m *segmenter) adjustPartDuration(du time.Duration) {
	if m.firstSegmentFinalized {
		return
	}

	// Avoid a crash by skipping invalid durations.
	if du == 0 {
		return
	}

	if _, ok := m.sampleDurations[du]; !ok {
		m.sampleDurations[du] = struct{}{}
		m.adjustedPartDuration = findCompatiblePartDuration(
			m.partDuration,
			m.sampleDurations,
		)
	}
}

func (m *segmenter) writeH264(ntp time.Time, pts time.Duration, nalus [][]byte) error {
	idrPresent := false
	nonIDRPresent := false

	for _, nalu := range nalus {
		typ := h264.NALUType(nalu[0] & 0x1F)
		switch typ {
		case h264.NALUTypeIDR:
			idrPresent = true

		case h264.NALUTypeNonIDR:
			nonIDRPresent = true
		}
	}

	if !idrPresent && !nonIDRPresent {
		return nil
	}

	avcc := h264.AVCCMarshal(nalus)

	var dts time.Duration
	if !m.videoFirstIDRReceived {
		// skip sample silently until we find one with an IDR
		if !idrPresent {
			return nil
		}

		m.videoFirstIDRReceived = true
		m.videoDTSExtractor = h264.NewDTSExtractor()
		m.videoSPS = m.videoTrack.SPS

		var err error
		dts, err = m.videoDTSExtractor.Extract(nalus, dts)
		if err != nil {
			return err
		}

		m.startDTS = dts
		dts = 0
		pts -= m.startDTS
	} else {
		var err error
		dts, err = m.videoDTSExtractor.Extract(nalus, pts)
		if err != nil {
			return err
		}

		pts -= m.startDTS
		dts -= m.startDTS
	}

	return m.writeH264Entry(ntp, &VideoSample{
		PTS:        m.muxerStartTime + int64(pts),
		DTS:        m.muxerStartTime + int64(dts),
		AVCC:       avcc,
		IdrPresent: idrPresent,
	})
}

func (m *segmenter) writeH264Entry(ntp time.Time, sample *VideoSample) error { //nolint:funlen
	sample, m.nextVideoSample = m.nextVideoSample, sample
	if sample == nil {
		return nil
	}

	next := m.nextVideoSample
	sample.NextDTS = next.DTS

	if m.currentSegment == nil {
		// create first segment
		m.currentSegment = newSegment(
			m.genSegmentID(),
			ntp,
			time.Duration(sample.DTS-m.muxerStartTime),
			m.muxerStartTime,
			m.segmentMaxSize,
			m.audioTrack,
			m.genPartID,
			m.onPartFinalized,
		)
	}

	m.adjustPartDuration(sample.duration())

	err := m.currentSegment.writeH264(sample, m.adjustedPartDuration)
	if err != nil {
		return err
	}

	if !next.IdrPresent {
		return nil
	}
	// switch segment
	sps := m.videoTrack.SPS
	spsChanged := !bytes.Equal(m.videoSPS, sps)

	if (time.Duration(sample.NextDTS-m.muxerStartTime)-
		m.currentSegment.startDTS) >= m.segmentDuration || spsChanged {
		err := m.currentSegment.finalize(next)
		if err != nil {
			return err
		}
		m.onSegmentFinalized(m.currentSegment)

		m.firstSegmentFinalized = true

		m.currentSegment = newSegment(
			m.genSegmentID(),
			ntp,
			time.Duration(next.PTS-m.muxerStartTime),
			m.muxerStartTime,
			m.segmentMaxSize,
			m.audioTrack,
			m.genPartID,
			m.onPartFinalized,
		)

		// if SPS changed, reset adjusted part duration
		if spsChanged {
			m.videoSPS = sps
			m.firstSegmentFinalized = false
			m.sampleDurations = make(map[time.Duration]struct{})
		}
	}

	return nil
}

func (m *segmenter) writeAAC(pts time.Duration, au []byte) error {
	return m.writeAACEntry(&AudioSample{
		PTS: int64(pts),
		AU:  au,
	})
}

func (m *segmenter) writeAACEntry(sample *AudioSample) error {
	// wait for the video track
	if !m.videoFirstIDRReceived {
		return nil
	}

	sample.PTS -= int64(m.startDTS)
	sample.PTS += m.muxerStartTime

	// put samples into a queue in order to
	// allow to compute the sample duration
	sample, m.nextAudioSample = m.nextAudioSample, sample
	if sample == nil {
		return nil
	}

	sample.NextPTS = m.nextAudioSample.PTS

	// wait for the video track
	if m.currentSegment == nil {
		return nil
	}

	err := m.currentSegment.writeAAC(sample)
	if err != nil {
		return err
	}

	return nil
}

File: src/hls/segmenter.rs
use crate::{
    error::{AdjustPartDurationError, CreateSegmenterError, SegmenterWriteH264Error},
    playlist::Playlist,
    segment::Segment,
    types::IdCounter,
};
use common::{
    time::{DurationH264, UnixNano, H264_MILLISECOND, H264_SECOND},
    H264Data, VideoSample,
};
use std::{collections::HashSet, sync::Arc};
use tokio_util::sync::DropGuard;

// Opaque wrapper around segmenter that will cancel the muxer when dropped.
pub struct H264Writer {
    segmenter: Segmenter,
    _guard: DropGuard,
}

impl H264Writer {
    #[must_use]
    pub fn new(segmenter: Segmenter, guard: DropGuard) -> Self {
        Self {
            segmenter,
            _guard: guard,
        }
    }
    pub async fn write_h264(&mut self, data: H264Data) -> Result<(), SegmenterWriteH264Error> {
        self.segmenter.write_h264(data).await
    }

    #[cfg(test)]
    #[allow(clippy::unwrap_used)]
    pub async fn test_write(&mut self, pts: i64, avcc: Vec<u8>, random_access: bool) {
        use common::time::DtsOffset;
        use sentryshot_padded_bytes::PaddedBytes;

        self.write_h264(H264Data {
            pts: common::time::UnixH264::new(pts),
            dts_offset: DtsOffset::new(0),
            avcc: Arc::new(PaddedBytes::new(avcc)),
            random_access_present: random_access,
        })
        .await
        .unwrap();
    }
}

pub struct Segmenter {
    segment_duration: DurationH264,
    part_duration: DurationH264,
    segment_max_size: u64,
    playlist: Arc<Playlist>,
    muxer_id: u16,

    muxer_start_time: UnixNano,
    //last_video_params: Vec<Vec<u8>>,
    current_segment: Option<Segment>,
    segment_id_counter: IdCounter,
    part_id_counter: IdCounter,
    next_sample: VideoSample,
    first_segment_finalized: bool,
    sample_durations: HashSet<DurationH264>,
    adjusted_part_duration: DurationH264,
}

impl Segmenter {
    pub fn new(
        segment_duration: DurationH264,
        part_duration: DurationH264,
        segment_max_size: u64,
        playlist: Arc<Playlist>,
        muxer_id: u16,
        muxer_start_time: UnixNano,
        first_sample: H264Data,
    ) -> Result<Self, CreateSegmenterError> {
        if !first_sample.random_access_present {
            return Err(CreateSegmenterError::NotIdr);
        }
        if *first_sample.dts_offset != 0 {
            return Err(CreateSegmenterError::DtsNotZero);
        }

        let next_sample = VideoSample {
            pts: muxer_start_time.into(),
            dts_offset: first_sample.dts_offset,
            avcc: first_sample.avcc,
            random_access_present: first_sample.random_access_present,
            duration: DurationH264::new(0),
        };

        Ok(Self {
            segment_duration,
            part_duration,
            segment_max_size,
            playlist,
            muxer_start_time,
            muxer_id,
            //last_video_params: Vec::new(),
            current_segment: None,
            segment_id_counter: IdCounter::new(7), // 7 required by iOS.
            part_id_counter: IdCounter::new(0),
            next_sample,
            first_segment_finalized: false,
            sample_durations: HashSet::new(),
            adjusted_part_duration: DurationH264::new(0),
        })
    }

    // iPhone iOS fails if part durations are less than 85% of maximum part duration.
    // find a part duration that is compatible with all received sample durations.
    fn adjust_part_duration(&mut self, du: DurationH264) -> Result<(), AdjustPartDurationError> {
        if self.first_segment_finalized {
            return Ok(());
        }

        // Skip invalid durations.
        if *du == 0 {
            return Ok(());
        }

        if !self.sample_durations.contains(&du) {
            self.sample_durations.insert(du);
            self.adjusted_part_duration =
                find_compatible_part_duration(self.part_duration, &self.sample_durations)
                    .ok_or(AdjustPartDurationError::Error)?;
        }
        Ok(())
    }

    pub async fn write_h264(&mut self, data: H264Data) -> Result<(), SegmenterWriteH264Error> {
        use crate::error::SegmenterWriteH264Error::*;

        let sample = VideoSample {
            pts: data.pts,
            dts_offset: data.dts_offset,
            avcc: data.avcc,
            random_access_present: data.random_access_present,
            duration: DurationH264::new(0),
        };

        let next_dts = sample.dts().ok_or(Dts)?;

        // Put samples in a queue in order to compute sample duration.
        let mut sample = std::mem::replace(&mut self.next_sample, sample);

        let sample_dts = sample.dts().ok_or(Dts)?;

        sample.duration = next_dts
            .checked_sub(sample_dts)
            .ok_or(ComputeSampleDuration)?
            .into();
        if *sample.duration < 0 {
            sample.duration = DurationH264::new(0);
        }

        self.adjust_part_duration(sample.duration)?;

        let current_segment = self.current_segment.get_or_insert_with(|| {
            Segment::new(
                self.segment_id_counter.next_id(),
                self.muxer_id,
                sample_dts,
                self.muxer_start_time,
                self.segment_max_size,
                self.playlist.clone(),
                &mut self.part_id_counter,
            )
        });

        let segment_start_dts = current_segment.start_dts();
        current_segment
            .write_h264(
                sample,
                self.adjusted_part_duration,
                &mut self.part_id_counter,
            )
            .await?;

        // switch segment
        if data.random_access_present {
            //videoParams := extractVideoParams(m.videoTrack)
            //paramsChanged := !videoParamsEqual(m.lastVideoParams, videoParams)

            if DurationH264::from(
                next_dts
                    .checked_sub(segment_start_dts)
                    .ok_or(SwitchSegment)?,
            ) >= self.segment_duration
            /*|| paramsChanged*/
            {
                let next_segment = Segment::new(
                    self.segment_id_counter.next_id(),
                    self.muxer_id,
                    sample_dts,
                    self.muxer_start_time,
                    self.segment_max_size,
                    self.playlist.clone(),
                    &mut self.part_id_counter,
                );
                let prev_segment = std::mem::replace(current_segment, next_segment);

                let Some(finalized_segment) = prev_segment.finalize(next_dts).await? else {
                    // Cancelled.
                    return Ok(());
                };
                self.playlist.on_segment_finalized(finalized_segment).await;

                self.first_segment_finalized = true;

                /*if paramsChanged {
                    m.lastVideoParams = videoParams
                    m.firstSegmentFinalized = false

                    // reset adjusted part duration
                    m.sampleDurations = make(map[time.Duration]struct{})
                }*/
            }
        }

        Ok(())
    }
}

/*
fn extractVideoParams(track: *gortsplib.TrackH264) ->  Vec<Vec<u8>> {
    params := make([][]byte, 2);
    params[0] = track.SafeSPS();
    params[1] = track.SafePPS();
    return params
}

fn videoParamsEqual(p1: Vec<Vec<u8>>, p2: Vec<Vec<u8>>)-> bool {
    if len(p1) != len(p2) {
        return true
    }

    for i, p := range p1 {
        if !bytes.Equal(p2[i], p) {
            return false
        }
    }
    return true
}
*/

fn part_duration_is_compatible(
    part_duration: DurationH264,
    sample_duration: DurationH264,
) -> Option<bool> {
    if sample_duration > part_duration {
        return Some(false);
    }

    let mut f = part_duration.checked_div(sample_duration)?;
    if !(part_duration.checked_rem(sample_duration)?).is_zero() {
        f = f.checked_add(DurationH264::new(1))?;
    }
    f = f.checked_mul(sample_duration)?;

    Some(
        part_duration
            > f.checked_mul(DurationH264::new(85))?
                .checked_div(DurationH264::new(100))?,
    )
}

fn part_duration_is_compatible_with_all(
    part_duration: DurationH264,
    sample_durations: &HashSet<DurationH264>,
) -> Option<bool> {
    for sd in sample_durations {
        if !part_duration_is_compatible(part_duration, *sd)? {
            return Some(false);
        }
    }
    Some(true)
}

fn find_compatible_part_duration(
    min_part_duration: DurationH264,
    sample_durations: &HashSet<DurationH264>,
) -> Option<DurationH264> {
    let mut i = min_part_duration;
    while *i < 5 * H264_SECOND {
        if part_duration_is_compatible_with_all(i, sample_durations)? {
            break;
        }
        i = i.checked_add(DurationH264::new(5 * H264_MILLISECOND))?;
    }
    Some(i)
}
